File: frontend/src/lib/signed-url.ts
import { createHmac, timingSafeEqual } from 'node:crypto'

const DEFAULT_EXPIRY = 4 * 60 * 60 // 4 hours

function normalizeMediaPath(filePath: string): string {
  const mediaRoot = '/app/media'
  let relative = filePath.trim().replace(/\\/g, '/')
  if (relative.startsWith(mediaRoot)) {
    relative = relative.slice(mediaRoot.length)
  } else if (relative.startsWith('/media/')) {
    relative = relative.slice('/media'.length)
  }

  const segments = relative.split('/').filter(Boolean)
  if (
    segments.length === 0 ||
    segments.some(
      (segment) =>
        segment === '.' ||
        segment === '..' ||
        segment.includes('?') ||
        segment.includes('#') ||
        [...segment].some((character) => {
          const codePoint = character.codePointAt(0) ?? 0
          return codePoint < 32 || codePoint === 127
        })
    )
  ) {
    throw new Error('Invalid media path')
  }
  return `/media/${segments.join('/')}`
}

function signingKey(): string {
  const key = process.env.INTERNAL_API_KEY || process.env.AUTH_SECRET
  if (!key && process.env.NODE_ENV === 'production') {
    throw new Error(
      'INTERNAL_API_KEY or AUTH_SECRET is required to sign media URLs'
    )
  }
  return key || 'sneepcut-dev-key'
}

export function signMediaUrl(
  filePath: string,
  expiresIn = DEFAULT_EXPIRY
): string {
  const mediaPath = normalizeMediaPath(filePath)
  const expires = Math.floor(Date.now() / 1000) + expiresIn
  const message = `${mediaPath}:${expires}`
  const sig = createHmac('sha256', signingKey())
    .update(message)
    .digest('hex')
    .slice(0, 32)

  return `${mediaPath}?expires=${expires}&sig=${sig}`
}

export function verifyMediaSignature(
  path: string,
  expires: string,
  sig: string
): boolean {
  const exp = Number.parseInt(expires, 10)
  if (Number.isNaN(exp) || Date.now() / 1000 > exp) return false

  const message = `${path}:${exp}`
  const expected = createHmac('sha256', signingKey())
    .update(message)
    .digest('hex')
    .slice(0, 32)

  try {
    return timingSafeEqual(Buffer.from(sig), Buffer.from(expected))
  } catch {
    return false
  }
}

function localMediaPath(storedUrl: string): string | null {
  if (storedUrl.startsWith('/media/')) {
    return storedUrl.split('?', 1)[0] ?? null
  }

  if (!/^https:\/\//i.test(storedUrl)) return null
  try {
    const url = new URL(storedUrl)
    return url.pathname.startsWith('/media/') &&
      url.searchParams.has('expires') &&
      url.searchParams.has('sig')
      ? url.pathname
      : null
  } catch {
    return null
  }
}

export function resolveMediaUrl(
  filePath: string | null | undefined,
  storedUrl: string | null | undefined
): string | null {
  const storedLocalPath = storedUrl ? localMediaPath(storedUrl) : null
  if (storedUrl && /^https:\/\//i.test(storedUrl) && !storedLocalPath) {
    return storedUrl
  }
  if (filePath) {
    try {
      return signMediaUrl(filePath)
    } catch {
      return null
    }
  }

  if (storedLocalPath) {
    try {
      return signMediaUrl(storedLocalPath)
    } catch {
      return null
    }
  }
  return null
}

File: backend/app/utils/signed_url.py
from __future__ import annotations

import hashlib
import hmac
import time
from urllib.parse import parse_qs, urlencode, urlparse, urlunparse

from app.config import settings

DEFAULT_EXPIRY = 3600 * 4  # 4 hours


def _signing_key() -> bytes:
    secret = settings.internal_api_key or settings.nextauth_secret or "clipforge-dev-key"
    return secret.encode()


def sign_url(path: str, expires_in: int = DEFAULT_EXPIRY) -> str:
    expires = int(time.time()) + expires_in
    message = f"{path}:{expires}"
    sig = hmac.new(_signing_key(), message.encode(), hashlib.sha256).hexdigest()[:32]
    return f"{path}?expires={expires}&sig={sig}"


def verify_signature(path: str, expires: str, sig: str) -> bool:
    try:
        exp = int(expires)
    except (ValueError, TypeError):
        return False

    if time.time() > exp:
        return False

    message = f"{path}:{exp}"
    expected = hmac.new(_signing_key(), message.encode(), hashlib.sha256).hexdigest()[:32]
    return hmac.compare_digest(sig, expected)


def make_signed_media_url(file_path: str, base_url: str | None = None) -> str:
    if not file_path:
        return ""
    media_root = settings.local_media_root
    if file_path.startswith(media_root):
        relative = file_path[len(media_root):]
    else:
        relative = file_path

    if not relative.startswith("/"):
        relative = f"/{relative}"

    media_path = f"/media{relative}"
    signed = sign_url(media_path)
    url_base = base_url or settings.app_url
    return f"{url_base}{signed}"
